src/node/txorphanage.cpp

```diff
@@ -83,373 +174,487 @@ class TxOrphanageImpl final : public TxOrphanage {
-    while (!work_set.empty()) {
-        Wtxid wtxid = *work_set.begin();
-        work_set.erase(work_set.begin());
-
-        const auto orphan_it = m_orphans.find(wtxid);
-        if (orphan_it != m_orphans.end()) {
-            return orphan_it->second.tx;
-        }
+    auto it = m_orphans.get<ByPeer>().lower_bound(ByPeerView{peer, true, 0});
+    if (it != m_orphans.get<ByPeer>().end() && it->m_announcer == peer && it->m_reconsider) {
+        // Flip m_reconsider. Even if this transaction stays in orphanage, it shouldn't be
+        // reconsidered again until there is a new reason to do so.
+        static constexpr auto mark_reconsidered_modifier = [](auto& ann) { ann.m_reconsider = false; };
+        m_orphans.get<ByPeer>().modify(it, mark_reconsidered_modifier);
+        return it->m_tx;
     }
     return nullptr;
 }
 
+/** Return whether there is a tx that can be reconsidered. */
 bool TxOrphanageImpl::HaveTxToReconsider(NodeId peer)
 {
-    auto peer_it = m_peer_orphanage_info.find(peer);
-    if (peer_it == m_peer_orphanage_info.end()) return false;
-
-    auto& work_set = peer_it->second.m_work_set;
-    return !work_set.empty();
+    auto it = m_orphans.get<ByPeer>().lower_bound(ByPeerView{peer, true, 0});
+    return it != m_orphans.get<ByPeer>().end() && it->m_announcer == peer && it->m_reconsider;
 }
-
 void TxOrphanageImpl::EraseForBlock(const CBlock& block)
 {
-    std::vector<Wtxid> vOrphanErase;
-
+    std::set<Wtxid> wtxids_to_erase;
     for (const CTransactionRef& ptx : block.vtx) {
-        const CTransaction& tx = *ptx;
+        const CTransaction& block_tx = *ptx;
 
         // Which orphan pool entries must we evict?
-        for (const auto& txin : tx.vin) {
-            auto itByPrev = m_outpoint_to_orphan_it.find(txin.prevout);
-            if (itByPrev == m_outpoint_to_orphan_it.end()) continue;
-            for (auto mi = itByPrev->second.begin(); mi != itByPrev->second.end(); ++mi) {
-                const CTransaction& orphanTx = *(*mi)->second.tx;
-                vOrphanErase.push_back(orphanTx.GetWitnessHash());
+        for (const auto& input : block_tx.vin) {
+            auto it_prev = m_outpoint_to_orphan_it.find(input.prevout);
+            if (it_prev != m_outpoint_to_orphan_it.end()) {
+                // Copy all wtxids to wtxids_to_erase.
+                std::copy(it_prev->second.cbegin(), it_prev->second.cend(), std::inserter(wtxids_to_erase, wtxids_to_erase.end()));
             }
         }
     }
 
-    // Erase orphan transactions included or precluded by this block
-    if (vOrphanErase.size()) {
-        int nErased = 0;
-        for (const auto& orphanHash : vOrphanErase) {
-            nErased += EraseTx(orphanHash);
-        }
-        LogDebug(BCLog::TXPACKAGES, "Erased %d orphan transaction(s) included or conflicted by block\n", nErased);
+    unsigned int num_erased{0};
+    for (const auto& wtxid : wtxids_to_erase) {
+        num_erased += EraseTx(wtxid);
```

Comment:
in a1bdebf370bf2fc43c0204ca2d6291e73fc7f91f: nit: since the previous commit 42e59cd5a4a119ba15991178f0438b4a7ffb5bab, this is now an implicit bool-to-int-cast
```suggestion
        num_erased += EraseTx(wtxid) ? 1 : 0;
```
(not sure how much we care about those, I guess there are countless other similar instances in the codebase; might be interesting to try https://clang.llvm.org/extra/clang-tidy/checks/readability/implicit-bool-conversion.html one day)